File: server/utils/groqAgent.js
/**
 * Groq AI Agent
 * Uses Groq API (Llama 3.3 70B) for fast, expert soil analysis narrative
 * Falls back gracefully if API key is missing or quota exceeded
 */
const Groq = require('groq-sdk');
const crypto = require('crypto');

// Initialise client once at module level
const _apiKey = process.env.GROQ_API_KEY || '';
const groq = _apiKey ? new Groq({ apiKey: _apiKey }) : null;

function buildPrompt(data, engineResult, imageAnalysis) {
  const { healthScore, nutrientAnalysis, cropSuitability, yieldPrediction } = engineResult;

  const visionBlock = imageAnalysis ? `
## VISUAL SOIL ANALYSIS (AI Image Detection)
- Soil Colour: ${imageAnalysis.soilColour || 'N/A'}
- Texture: ${imageAnalysis.texture || 'N/A'}
- Moisture Level: ${imageAnalysis.moisture || 'N/A'}
- Organic Matter Visibility: ${imageAnalysis.organicMatter || 'N/A'}
- Estimated Soil Type: ${imageAnalysis.estimatedSoilType || 'N/A'}
- Surface Condition: ${imageAnalysis.surfaceCondition || 'N/A'}
- Visual Observations: ${imageAnalysis.visualObservations || 'N/A'}
` : '';

  return `You are an expert agricultural soil scientist and agronomist with 20+ years of experience in Indian farming.

A farmer has submitted their soil test data along with a soil image. Analyse it and respond ONLY with a valid JSON object — no markdown, no explanation, no code fences, just raw JSON.

## FARMER & FARM DETAILS
- Farmer: ${data.farmerName}, ${data.village}, ${data.district}, ${data.state}
- Farm Size: ${data.farmSize} acres | Crop: ${data.crop} | Previous Crop: ${data.previousCrop}
- Season: ${data.season} | Irrigation: ${data.irrigationType}
- Soil Type: ${data.soilType} | Soil Colour: ${data.soilColour}
${visionBlock}
## SOIL TEST RESULTS (User Entered)
- pH: ${data.soilPH} | EC: ${data.electricalConductivity} dS/m | Organic Carbon: ${data.organicCarbon}%
- Nitrogen: ${data.nitrogen} kg/ha | Phosphorus: ${data.phosphorus} kg/ha | Potassium: ${data.potassium} kg/ha

## WEATHER
- Temperature: ${data.temperature}°C | Humidity: ${data.humidity}% | Rainfall: ${data.rainfall} mm/year

## PRE-COMPUTED DATA
- Health Score: ${healthScore.overall}/100 (${healthScore.grade})
- Suitable Crops: ${cropSuitability.suitable.map(c => c.crop).join(', ') || 'None'}
- Issues: ${nutrientAnalysis.issues.map(i => `${i.param} ${i.level}`).join(', ') || 'None'}
- Predicted Yield: ${yieldPrediction.predicted}
${data.notes ? `- Farmer Notes: ${data.notes}` : ''}

Respond with this exact JSON structure:
{
  "soilStory": "3-4 sentence plain English summary of soil condition and most important action",
  "expertVerdict": "One powerful diagnosis sentence like a doctor giving verdict",
  "nutrientExplanation": {
    "nitrogen": "simple explanation of nitrogen level and crop impact",
    "phosphorus": "simple explanation of phosphorus level",
    "potassium": "simple explanation of potassium level",
    "ph": "what this pH means for nutrient availability",
    "organicCarbon": "organic carbon status and importance"
  },
  "cropAdvice": [
    { "crop": "name", "why": "why it suits this soil", "expectedYield": "yield estimate", "keyTip": "top tip for this crop" }
  ],
  "fertilizerNarrative": "2-3 sentences on fertilizer strategy in farmer-friendly language",
  "irrigationAdvice": "2 sentences on irrigation based on soil, crop and rainfall",
  "seasonalCalendar": [
    { "month": "month or range", "task": "farming task", "reason": "why this month" }
  ],
  "organicFarmingPath": "2-3 sentences on improving organic farming for this farm",
  "governmentSchemes": [
    { "scheme": "scheme name", "benefit": "what it provides", "eligibility": "who can apply" }
  ],
  "diseaseRisks": [
    { "disease": "disease or pest", "risk": "High/Medium/Low", "prevention": "prevention tip" }
  ],
  "marketInsight": "2 sentences on market demand for recommended crops in this region",
  "tamilSummary": "3-4 sentence summary in Tamil language for the farmer",
  "motivationalMessage": "One warm encouraging sentence for the farmer"
}`;
}

async function getGroqAnalysis(data, engineResult, imageAnalysis = null) {
  const placeholder = 'your_groq_api_key_here';
  // Timing-safe check: no key, or key matches placeholder
  const noKey = !_apiKey;
  const isPlaceholder = _apiKey.length === placeholder.length &&
    crypto.timingSafeEqual(Buffer.from(_apiKey), Buffer.from(placeholder));

  if (noKey || isPlaceholder || !groq) {
    return { available: false, reason: 'Groq API key not configured. Add GROQ_API_KEY to server/.env' };
  }

  try {
    const completion = await groq.chat.completions.create({
      model: 'llama-3.3-70b-versatile',
      messages: [
        {
          role: 'system',
          content: 'You are an expert agricultural soil scientist. Always respond with valid raw JSON only — no markdown, no code blocks, no extra text.'
        },
        {
          role: 'user',
      content: buildPrompt(data, engineResult, imageAnalysis)
        }
      ],
      temperature: 0.3,
      max_tokens: 2048,
      response_format: { type: 'json_object' },
    });

    const text = completion.choices[0]?.message?.content?.trim();
    const parsed = JSON.parse(text);

    return { available: true, model: 'llama-3.3-70b-versatile', ...parsed };
  } catch (err) {
    console.error('Groq API error:', err.message);
    return { available: false, reason: err.message };
  }
}

module.exports = { getGroqAnalysis };

File: backend/app/agents/groq_agent.py
"""
Groq AI Agent — builds narrative from image analysis results.
"""
import json
import logging
import httpx
from app.config import settings

logger = logging.getLogger(__name__)
GROQ_API_URL = "https://api.groq.com/openai/v1/chat/completions"
MODEL = "llama-3.3-70b-versatile"


def _build_prompt(soil_data: dict, engine_result: dict, user_name: str) -> str:
    hs = engine_result["healthScore"]
    cs = engine_result["cropSuitability"]
    yp = engine_result["yieldPrediction"]
    na = engine_result["nutrientAnalysis"]

    suitable_crops = ", ".join(c["crop"] for c in cs["suitable"]) or "None"
    issues_str     = ", ".join(f"{i['param']} {i['level']}" for i in na["issues"]) or "None"

    return f"""You are an expert agricultural soil scientist with 20+ years of experience in Indian farming.

A farmer named {user_name} has submitted their soil test results for analysis.

## FARMER & FARM DETAILS
- Farmer: {soil_data.get('farmerName', user_name)}
- Location: {soil_data.get('village', '')}, {soil_data.get('district', '')}, {soil_data.get('state', '')}
- Farm Size: {soil_data.get('farmSize', 1)} acres
- Crop: {soil_data.get('crop', 'Unknown')}
- Previous Crop: {soil_data.get('previousCrop', 'Unknown')}
- Season: {soil_data.get('season', 'Kharif')}
- Irrigation: {soil_data.get('irrigationType', 'Drip')}

## SOIL TEST PARAMETERS (Lab Measured)
- Soil Type: {soil_data.get('soilType', 'Unknown')}
- Soil Colour: {soil_data.get('soilColour', 'Unknown')}
- pH: {soil_data.get('soilPH', 7.0)}
- EC: {soil_data.get('electricalConductivity', 0.5)} dS/m
- Organic Carbon: {soil_data.get('organicCarbon', 0.75)}%
- Nitrogen: {soil_data.get('nitrogen', 80)} kg/ha
- Phosphorus: {soil_data.get('phosphorus', 40)} kg/ha
- Potassium: {soil_data.get('potassium', 150)} kg/ha
- Temperature: {soil_data.get('temperature', 28)}°C
- Humidity: {soil_data.get('humidity', 65)}%
- Rainfall: {soil_data.get('rainfall', 800)} mm/year

## PRE-COMPUTED ENGINE RESULTS
- Health Score: {hs['overall']}/100 ({hs['grade']})
- Suitable Crops: {suitable_crops}
- Issues: {issues_str}
- Predicted Yield: {yp['predicted']}

Respond ONLY with valid raw JSON:
{{
  "soilStory": "3-4 sentence plain English summary",
  "expertVerdict": "One powerful diagnosis sentence",
  "nutrientExplanation": {{
    "nitrogen": "explanation",
    "phosphorus": "explanation",
    "potassium": "explanation",
    "ph": "pH explanation",
    "organicCarbon": "OC explanation"
  }},
  "cropAdvice": [
    {{"crop": "name", "why": "reason", "expectedYield": "estimate", "keyTip": "tip"}}
  ],
  "fertilizerNarrative": "2-3 sentences",
  "irrigationAdvice": "2 sentences",
  "seasonalCalendar": [
    {{"month": "month", "task": "task", "reason": "reason"}}
  ],
  "organicFarmingPath": "2-3 sentences",
  "diseaseRisks": [
    {{"disease": "name", "risk": "High/Medium/Low", "prevention": "tip"}}
  ],
  "marketInsight": "2 sentences",
  "tamilSummary": "3-4 sentences in Tamil language summarising soil health, best crop and key advice for the farmer",
  "motivationalMessage": "One warm encouraging sentence"
}}"""


async def get_groq_analysis(soil_data: dict, engine_result: dict, user_name: str = "Farmer") -> dict:
    api_key = settings.GROQ_API_KEY
    if not api_key or api_key == "your_groq_api_key_here":
        return {"available": False, "reason": "Groq API key not configured"}

    payload = {
        "model": MODEL,
        "messages": [
            {"role": "system", "content": "You are an expert agricultural soil scientist. Always respond with valid raw JSON only."},
            {"role": "user", "content": _build_prompt(soil_data, engine_result, user_name)},
        ],
        "temperature": 0.3,
        "max_tokens": 2048,
        "response_format": {"type": "json_object"},
    }

    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.post(
                GROQ_API_URL,
                headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
                json=payload,
            )
            resp.raise_for_status()
            text = resp.json()["choices"][0]["message"]["content"].strip()
            parsed = json.loads(text)
            return {"available": True, "model": MODEL, **parsed}
    except Exception as e:
        logger.error(f"Groq API error: {e}")
        return {"available": False, "reason": str(e)}
